### PR TITLE
Feature more filter funtions revised

### PR DESCRIPTION
**Pull request by cKlee**
We discussed this at https://github.com/LinkedDataFragments/Client.js/issues/22. Code is from the feature-more-filter-funtions branch. I just updated the deprecated N3Util.isUri to N3Util.isIRI. 381 test were passing.
Although the @license comments may be wrong.

### CODE REVIEW

**Comment by RubenVerborgh** [comment]
Thanks for this effort!

I'm a bit concerned about the missing history, but I can probably rebase that; the linting would still need to be fixed, however.

**Comment by RubenVerborgh** [comment]
Thanks for these fixes!

File: lib/util/SparqlExpressionEvaluator.js [hunk 1]
```diff
@@ -1,17 +1,44 @@
-/*! @license MIT ©2014-2016 Ruben Verborgh, Ghent University - imec */
+/*! @license MIT ©2014-2016 Ruben Verborgh, Miel Vander Sande, Ghent University - imec */
 
 var N3Util = require('n3').Util,
-    createErrorType = require('./CustomError');
+    UUID = require('uuid'),
+    crypto = require('cryptojs');
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: TODO: check the (size) impact of these libraries on the browser version.

File: lib/util/SparqlExpressionEvaluator.js [hunk 2]
```diff
@@ -1,17 +1,44 @@
...
+  'byte',
+  'nonNegativeInteger',
+  'unsignedLong',
+  'unsignedInt',
+  'unsignedShort',
+  'unsignedByte',
+  'positiveInteger',
+].map(function (type) {
+  return XSD + type;
+});
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: TODO probably inline this a bit, spreading the array over 2 or 3 lines.

Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: TODO This should be a hash, not an array, since it is only used for membership checks (`indexOf` has linear time, `in` has constant time).

File: lib/util/SparqlExpressionEvaluator.js [hunk 3]
```diff
@@ -23,7 +50,7 @@ function SparqlExpressionEvaluator(expression) {
   if (!expression) return noop;
   var expressionType = expression && expression.type || typeof expression,
       evaluator = evaluators[expressionType];
-  if (!evaluator) throw new UnsupportedExpressionError(expressionType);
+  if (!evaluator) throw new Error('Unsupported expression type: ' + expressionType);
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: @cKlee What's the rationale behind this change?

File: lib/util/SparqlExpressionEvaluator.js [hunk 4]
```diff
@@ -32,34 +59,53 @@ SparqlExpressionEvaluator.evaluate = function (expression, bindings) {
...
-// The null operation
-function noop() { }
-
 // Evaluators for each of the expression types
 evaluators = {
   // Does nothing
-  null: function () { return noop; },
+  null: function () {
+    return noop;
+  },
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: What do we need this for again?

Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: Also, was the reformatting necessary (did the linter complain)?

File: lib/util/SparqlExpressionEvaluator.js [hunk 5]
```diff
@@ -32,34 +59,53 @@ SparqlExpressionEvaluator.evaluate = function (expression, bindings) {
...
+    var operatorName = expression.operator,
         operator = operators[operatorName];
     if (!operator)
-      throw new UnsupportedOperatorError(operatorName);
-    if (operator.length !== expression.args.length)
-      throw new InvalidArgumentsNumberError(operatorName, expression.args.length, operator.length);
-
+      throw new Error('Unsupported operator: ' + operatorName.toUpperCase() + '.');
+    if (operator.argBounds) {
+      var min = operator.argBounds.min, max = operator.argBounds.max;
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: TODO: Perhaps we want `expression.args.length` also as a variable here.

**Comment by mielvds** [comment]
Nice job! Looks good, so the feature-more-filter-funtions branch can be deleted?

File: lib/util/SparqlExpressionEvaluator.js [hunk 6]
```diff
@@ -32,34 +59,53 @@ SparqlExpressionEvaluator.evaluate = function (expression, bindings) {
...
+      if ((min && max && (min > expression.args.length || max < expression.args.length)) ||
+        (min && (min > expression.args.length)) ||
+        (max && (max < expression.args.length))) {
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: Good that we have bounds checking.

Could it be that `max` is set to `0`, i.e., no arguments? That would fail the bound check. E.g., `UUID` and `strUUID`, if we pass them arguments, I think this check would not work.

Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: …Ah, but maybe then there wouldn't be `argBounds`, but just `length`.

File: lib/util/SparqlExpressionEvaluator.js [hunk 7]
```diff
@@ -101,121 +158,489 @@ evaluators = {
         switch (operator.resultType) {
         case 'numeric':
           // TODO: determine type instead of taking the type of the first argument
-          var type = N3Util.getLiteralType(origArgs[0]) || XSD_INTEGER;
+          var type = isNumeric(origArgs[0]) ? N3Util.getLiteralType(origArgs[0]) : XSD + 'integer';
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: Does this fully resolve the TODO above? If so, it can be removed.

File: lib/util/SparqlExpressionEvaluator.js [hunk 8]
```diff
@@ -101,121 +158,489 @@ evaluators = {
...
+  },
+  '!': function (a) {
+    return !a;
+  },
+  '&&': function (a, b) {
+    return a && b;
+  },
+  '||': function (a, b) {
+    return a || b;
+  },
```
Comment (RubenVerborgh) [inline lib/util/SparqlExpressionEvaluator.js]: TODO: where these inlinings necessary for the linter? Probably not, unless the linter changed. I'd inline them again, since this code now takes up a lot of space for what it does.

**Comment by mielvds** [comment]
Hmm, still some issues here: see comments on the code diff

**Comment by cKlee** [comment]
Sorry I erased them. I can put them back in in. Should I?

**Comment by RubenVerborgh** [comment]
@cKlee Yes please.

File: lib/util/SparqlExpressionEvaluator.js (hunk 3 above)
Comment (mielvds) [inline lib/util/SparqlExpressionEvaluator.js]: There isn't one, this was probably reverted by copy paste, since this branch existed  before https://github.com/LinkedDataFragments/Client.js/commit/73e690b3480784b9d13ef85634da1bdf1496c4ea

**Comment by RubenVerborgh** [comment]
@mielvds Given that this is based on your code, do you want to take ownership of this?

I would like to do the final merge though, since I want to fix a couple of things in the branching / commit history.

File: test/util/SparqlExpressionEvaluator-test.js [hunk 9]
```diff
@@ -21,474 +21,824 @@ describe('SparqlExpressionEvaluator', function () {
     describe('of a literal', function () {
       var evaluator = SparqlExpressionEvaluator('"abc"');
       it('should return the literal value', function () {
-        evaluator({ '?a': 'a' }).should.equal('"abc"');
+        evaluator({
+          '?a' : 'a',
```
Comment (RubenVerborgh) [inline test/util/SparqlExpressionEvaluator-test.js]: TODO: This should not have linted; we don't have spaces before our colons.

**Comment by RubenVerborgh** [comment]
@cKlee Thanks a lot for your work, we will merge this soon.

My main concern (apart from some detailed comments) is that the code style has changed a lot, I think unnecessarily in some places. I appreciate your work on having compatibility with our linter settings, but could it be that more was changed than was necessary? Did you apply automated corrections?

File: lib/util/SparqlExpressionEvaluator.js [hunk 10]
```diff
@@ -101,121 +158,489 @@ evaluators = {
...
+  'coalesce': function (a, b) {
+    return operators.bound.call(this, a) === XSD_TRUE ? this[a] : b;
+  },
+  'exists': function (a) {
+    throw new Error('EXISTS not yet supported');
+  },
+  'not exists': function (a) {
+    throw new Error('NOT EXISTS not yet supported');
+  },
```
Comment (mielvds) [inline lib/util/SparqlExpressionEvaluator.js]: correct, hence the error

**Comment by RubenVerborgh** [comment]
BTW First time I'm using this "review" function, so things might not fully go as expected.

**Comment by mielvds** [comment]
This review is still confusing yes :)

I could take ownership... but lots of the issues here are because this file was partially rewritten by you while the branch was asleep. So you'll need to do the final checks, also with the tests. I can do a cleanup in the meantime.

**Comment by RubenVerborgh** [comment]
@mielvds Perfect, deal!

**Comment by cKlee** [comment]
Let me know if I could do anything.

**Comment by mielvds** [comment]
Replaced by #30
